### PR TITLE
Allow custom VolumeClaim names in a StatefulSet

### PR DESCRIPTION
**Pull request by MilovdZee**
**What this PR does / why we need it**:
In our environment we have a department that manages the storage and also creates the VolumeClaims. Those names are preverably short and not prone to changes. StatefulSets use a template for the claims and generate a name like \<claimname\>-\<statefulSet name\>-\<ordinal\>. This means that if somebody likes to change the name of the statefulset the name of the claim should also change. As this happens a lot this creates a lot of work. Allowing for more stable PVC names helps in managing the cluster.
The proposal is to use \<baseName\>-\<ordinal\> where the baseName is user defined.

For example:
```
  volumeClaimTemplates: 
  - metadata: 
      name: myClaim
      annotations: 
        volumeClaimBaseName: storageClaim 
    spec: 
      accessModes: [ "ReadWriteMany" ] 

```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Allow custom VolumeClaim name in StatefulSet volumeClaimTemplates
```

### CODE REVIEW

**Comment by MilovdZee** [comment]
CLA signed

**Comment by MilovdZee** [comment]
CLA Signed and edited github e-mail address to match commit e-mail

**Comment by MilovdZee** [comment]
Hello,

@foxish and @janetkuo: Do I need to fix something or will this be merged with the following release.  "Submit Queue — PR does not have lgtm label." is not completely clear for me.

Thanks,
Milo

**Comment by foxish** [comment]
Needs LGTM and approval. @kow3ns PTAL.

**Comment by MilovdZee** [comment]
And these LGTM and approval are both things I can't do I assume? It does LGTM :)

**Comment by foxish** [comment]
@MilovdZee, yes, needs a review here. @msau42 PTAL also - wrt stable generated PVC names.

**Comment by msau42** [comment]
Just checked the dynamic provisioning volume spreading code for statefulsets [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util.go#L386).  With this change, it will use `volumeClaimBaseName` as the hash, which is fine if you only have one PVC statefulset pod, but if you have two with two different base names, then the hashes will be different and the volumes for a single pod could be provisioned in different zones.

That doesn't matter for this particular use case, where the volumes are preprovisioned, but if people wanted to use this naming feature for dynamic provisioning as well, then this particular scenario needs to be documented as not supported.

**Comment by thomasliddledba** [comment]
This would be a helpful feature for a use case that I have?  Any timeframe on when this could be merged?

File: pkg/controller/statefulset/stateful_set_utils.go [hunk 1]
```diff
@@ -98,6 +98,11 @@ func getPodName(set *apps.StatefulSet, ordinal int) string {
 // must be a PersistentVolumeClaim from set's VolumeClaims template.
 func getPersistentVolumeClaimName(set *apps.StatefulSet, claim *v1.PersistentVolumeClaim, ordinal int) string {
 	// NOTE: This name format is used by the heuristics for zone spreading in ChooseZoneForVolume
+	var volumeClaimBaseName string
+	volumeClaimBaseName = claim.ObjectMeta.Annotations["volumeClaimBaseName"]
```
Comment (janetkuo) [inline pkg/controller/statefulset/stateful_set_utils.go]: What happens when this annotation is changed after the StatefulSet and its pods are created? Volume claim names should be immutable.

**Comment by janetkuo** [comment]
@MilovdZee FYI the PR submit process: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#the-pr-submit-process

**Pull request closed without merging**